java/arcs/core/host/AbstractArcHost.kt

```diff
@@ -10,43 +10,239 @@
  */
 package arcs.core.host
 
+import arcs.core.data.CollectionType
 import arcs.core.data.Plan
-import kotlin.reflect.KClass
+import arcs.core.data.SingletonType
+import arcs.core.storage.api.Handle
+import arcs.core.storage.handle.HandleManager
+import arcs.core.util.Time
+
+typealias ParticleConstructor = suspend () -> Particle
+typealias ParticleRegistration = Pair<ParticleIdentifier, ParticleConstructor>
 
 /**
  * Base helper class for [ArcHost] implementations to provide implementation of
  * registration.
+ *
+ * @property initialParticles The initial set of [Particle]s that this host contains.
  */
-abstract class AbstractArcHost(
-    private var particles: MutableList<ParticleIdentifier> = mutableListOf()
-) : ArcHost {
-    // TODO: fix this, qualifiedName is not supported on JS
+abstract class AbstractArcHost(vararg initialParticles: ParticleRegistration) : ArcHost {
+    private val particles: MutableMap<ParticleIdentifier, ParticleConstructor> = mutableMapOf()
+    private val runningArcs: MutableMap<String, ArcHostContext> = mutableMapOf()
+
+    init {
+        initialParticles.toList().associateByTo(particles, { it.first }, { it.second })
+    }
+
     override val hostId = this::class.className()
 
-    protected fun registerParticle(particle: ParticleIdentifier) {
-        particles.add(particle)
+    protected fun registerParticle(particle: ParticleIdentifier, constructor: ParticleConstructor) {
+        particles.put(particle, constructor)
     }
 
     protected fun unregisterParticle(particle: ParticleIdentifier) {
         particles.remove(particle)
     }
 
-    override suspend fun registeredParticles(): List<ParticleIdentifier> = particles
+    override suspend fun registeredParticles(): List<ParticleIdentifier> = particles.keys.toList()
+
+    // VisibleForTesting
+    protected fun getArcHostContext(arcId: String) = runningArcs[arcId]
+
+    /** Subclasses may override this to load persistent context state. */
+    protected suspend fun lookupOrCreateArcHostContext(partition: Plan.Partition): ArcHostContext =
+        runningArcs[partition.arcId] ?: ArcHostContext()
+
+    /** Subclasses may override this to store persistent context state. */
+    protected suspend fun updateArcHostContext(arcId: String, context: ArcHostContext) {
```

Comment:
It's basically the way to persist the state, subclasses would persist into storage instead of a memory cache, so that if a service crashes and is restarted, it can recover the previous state.

Reply:
Cool, that's good context to provide in the kdoc.